sdk/core/azure-core/src/main/java/com/azure/core/util/SharedExecutorService.java

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.implementation.ReflectionUtils;
+import com.azure.core.implementation.ReflectiveInvoker;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * An {@link ExecutorService} that is shared by multiple consumers.
+ * <p>
+ * The shared executor service is created using the following configuration settings:
+ * <ul>
+ *     <li>{@code azure.sdk.sharedpool.size} - The size of the shared executor service. If not set, it defaults to 10
+ *     times the number of available processors.</li>
+ *     <li>{@code azure.sdk.sharedpool.keepalivemillis} - The keep alive time for threads in the shared executor
+ *     service. If not set, it defaults to 60 seconds.</li>
+ *     <li>{@code azure.sdk.sharedpool.virtual} - A boolean flag to indicate if the shared executor service should use
+ *     virtual threads. If not set, it defaults to true. Ignored if virtual threads are not available in the
+ *     runtime.</li>
+ * </ul>
+ */
+@SuppressWarnings({ "resource", "NullableProblems" })
+public final class SharedExecutorService implements ExecutorService {
+    private static final ClientLogger LOGGER = new ClientLogger(SharedExecutorService.class);
+
+    private static final SharedExecutorService INSTANCE = new SharedExecutorService();
+
+    // Shared thread counter for all instances of SharedExecutorService created using the empty factory method.
+    private static final AtomicLong AZURE_SDK_THREAD_COUNTER = new AtomicLong();
+    private static final String AZURE_SDK_THREAD_NAME = "azure-sdk-global-thread-";
+
+    // The thread pool size for the shared executor service.
+    //
+    // This uses the configuration setting 'azure.sdk.sharedpool.size' if set, otherwise it defaults to 10 times the
+    // number of available processors.
+    // If 'azure.sdk.sharedpool.size' is set to a non-integer, negative value, or zero, the default value is used.
+    private static final int THREAD_POOL_SIZE
+        = getConfig("azure.sdk.sharedpool.size", Integer::parseInt, 10 * Runtime.getRuntime().availableProcessors());
```

Comment:
Let's use `threadpool` or `shared.threadpool` to not confuse it with other types of shared pooling we may have.